components/tikv_util/src/trend.rs

```diff
@@ -0,0 +1,737 @@
+        } else {
+            if self.previous_window.is_overflow() {
+                self.previous_window.std_ev()
+            } else {
+                0.0
+            }
+        };
+        f64::max(margin_error, self.margin_error_base)
+    }
+
+    #[inline]
+    // Return bool: shoud_skip_current_value
+    fn try_flip(&mut self, value: f64, now: Instant, increasing_rate: f64) -> bool {
+        if !self.current_window.is_overflow() {
+            return false;
+        }
+        let current_avg = self.current_window.avg();
+        let margin_error = self.margin_error();
+
+        // The output margin_error can up to `self.flip_margin_error_multiple + 1`
+        // without flipping (increasing_rate already minus a margin_error)
+        let flip_margin_error = margin_error * self.flip_margin_error_multiple;
+        let delta = f64::abs(value - current_avg);
+
+        // Strict condition for exiting flipping (to do actual flipping)
+        if self.flipping_start_time.is_some() {
+            // Make sure not stuck at flipping phase by using `time_based_multiple`,
+            // increase by time
+            //   - Expectation of time_based_multiple: starts at 0.0, to `margin_error * 5%`
+            //     at 4min, to 10% at 12min, to 20% at 28min
+            //   - f64::abs() is for preventing crash in case the server time is adjusted
+            let flipping_duration = now.duration_since(self.flipping_start_time.unwrap());
+            let time_based_multiple =
+                (f64::abs(flipping_duration.as_secs() as f64) / 240.0 + 1.0).log2() / 20.0;
+            if f64::abs(increasing_rate) > margin_error * time_based_multiple {
+                // Keep flipping, skip the huge-changing phase, wait for stable
+                info!(
+                    "history window flipping: continue";
+                    "name" => self.name,
+                    "delta" => delta,
+                    "flip_margin_error" => margin_error,
+                    "time_based_multiple" => time_based_multiple,
+                    "increasing_rate" => increasing_rate,
+                    "flipping_duration" => flipping_duration.as_secs(),
+                );
+                return true;
+            } else {
+                // The huge-changing phase ends, do flipping
+                self.flip();
+                self.flipping_start_time = None;
+                self.last_flipped_time = now;
+                info!(
+                    "history window flipping: end";
+                    "name" => self.name,
+                    "delta" => delta,
+                    "flip_margin_error" => margin_error,
+                    "time_based_multiple" => time_based_multiple,
+                    "increasing_rate" => increasing_rate,
+                    "flipping_duration" => flipping_duration.as_secs(),
+                );
+                return false;
+            }
+        }
+
+        // Loose condition for entering flipping
```

Comment:
Loose for `enter` still very strict but just looser comparing to `exit`:
* A bit more flipping is still fairly safe

Strict for `exit` can skip the whole `increasing`(or decreasing) phase, we don't need any little part of the changing phase's data get into this history window, because the data in history should be the **normal values**